### PR TITLE
Make logging optional while decoding files

### PR DESCRIPTION
**Pull request by matteodelabre**
Currently, the notebook decoder outputs various debugging information to the standard output while decoding files. While this output can be useful at times, it is, in my opinion, unnecessary for most users of the library. The proposed changes include:

* moving logging instructions behind a conditionally-compiled macro;
* adding a disabled-by-default CMake flag allowing users to manually turn on logging when needed;
* when disabled, compiling the logging macro to a noop, so as to not affect normal usage performance;
* when enabled, logging to the standard error stream instead of the standard output. If some parts of the library are further extended and need to output non-debugging information to the standard output, users that have enabled logging would still be able to distinguish the two kinds of information.

### CODE REVIEW

**Comment by ax3l** [comment]
Thank you for the PR!
The changes look good, can you please instead of a compile-time option just add a runtime option? We could either put it through with a parameter or by reading an environment var for activation.

I am aware about the (minimal) runtime costs, but if you say we should avoid them altogether we can just remove my log outputs and add a little executable that outputs the same noisy details. I just used those during research.

**Comment by matteodelabre** [comment]
Thanks for the feedback. I agree that a runtime option would be more convenient than a compile-time option. However, as you mentioned, it would create a small runtime cost for end-users that do not care about the logging instructions.

For these reasons, I think creating a separate executable, as you proposed, would be the best tradeoff, especially given that this information is most likely to be used only internally, while debugging reading problems.

I set up an executable called `dump` in the latest commit that does just that. Let me know what you think.

File: include/rmlab/tool/dump.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/* Copyright 2017-2018 Axel Huebl
```
Comment (ax3l) [inline include/rmlab/tool/dump.cpp]: Please feel free to add your name to this line :)
(Is "pure ASCII" ok for you? Had bad experiences with some compilers with non-ascii comments in source files.)

File: include/rmlab/tool/dump.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            std::size_t line_id = 0;
+
+            for( const auto & line : layer.lines )
+            {
+                std::cout << "    line " << line_id
+                          << " ---------------------\n"
+                          << "      brush type: " << line.brush_type << '\n'
+                          << "      color int32: " << line.color << '\n'
+                          << "      magic 4byte: "
+                                << line.unknown_line_attribute << '\n'
```
Comment (ax3l) [inline include/rmlab/tool/dump.cpp]: is this formatting intentional?

File: include/rmlab/tool/dump.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                          << "      brush size: "
+                                << line.brush_base_size << '\n'
+                          << "      no. points: " << line.points.size() << '\n';
+
+                std::size_t point_id = 0;
```
Comment (ax3l) [inline include/rmlab/tool/dump.cpp]: `0u` ;)

File: include/rmlab/tool/dump.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+        std::size_t layer_id = 0;
+
+        for( const auto & layer : page.layers )
+        {
+            std::cout << "  layer " << layer_id
+                      << " ----------------------\n"
+                      << "    no. of lines: " << layer.lines.size() << '\n';
+
+            std::size_t line_id = 0;
```
Comment (ax3l) [inline include/rmlab/tool/dump.cpp]: `0u` ;)

File: include/rmlab/tool/dump.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    std::cout << "no. of pages: " << notebook.pages.size() << '\n';
+    std::size_t page_id = 0;
+
+    for( const auto & page : notebook.pages )
+    {
+        std::cout << "page " << page_id
+                  << " -------------------------\n"
+                  << "  no. of layers: " << page.layers.size() << '\n';
+
+        std::size_t layer_id = 0;
```
Comment (ax3l) [inline include/rmlab/tool/dump.cpp]: `0u` ;)

File: include/rmlab/tool/dump.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    if( argc != 2 )
+    {
+        std::cerr << "Usage: " << argv[0] << " path/to/filePrefix" << std::endl;
+        return 1;
+    }
+
+    rmlab::Notebook notebook( argv[1] );
+
+    std::cout << "no. of pages: " << notebook.pages.size() << '\n';
+    std::size_t page_id = 0;
```
Comment (ax3l) [inline include/rmlab/tool/dump.cpp]: `0u` ;)

File: include/rmlab/tool/dump.cpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+                ++line_id;
+            }
+
+            ++layer_id;
+        }
+
+        ++page_id;
+    }
+}
```
Comment (ax3l) [inline include/rmlab/tool/dump.cpp]: Maybe add `return 0;`?

**Review by ax3l — COMMENTED** [review]
Can you please also remove the `std::cout` in `Notebook::save`?

**Comment by matteodelabre** [comment]
All done! Regarding the formatting, the intent was to visually group together lines in the code that output a single line in the terminal, but it turns out it is not that readable.

File: include/rmlab/tool/dump.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
+/* Copyright 2017-2018 Axel Huebl, Matteo Delabre
+ *
+ * This file is part of lines-are-beautiful.
+ *
+ * lines-are-beautiful is free software: you can edistribute it and/or modify
```
Comment (ax3l) [inline include/rmlab/tool/dump.cpp]: +r :)

File: CMakeLists.txt [hunk 9]
```diff
@@ -106,6 +106,13 @@ else()
     message(STATUS "PNGwriter NOT found! lines2png will not be build!")
 endif()
 
+add_executable(dump
```
Comment (ax3l) [inline CMakeLists.txt]: there is a merge conflict here that needs rebasing :)

**Review by ax3l — APPROVED** [review]
Thank you! :sparkles:

**Pull request merged**